[Core] StateHash Interfaces & Diagrams

Olshansk opened the pull request:
## Description

Interface updates and diagrams to document the cross-module flow for stateHash implementation.

## Issue

Fixes part of #251

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [x] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

The goal of this change is to show the interface changes and plan of action to implement the state hash implementation.

Documentation and external interface changes required for updating the state hash. Details related to the internals of the persistence module will be done in a follow up commit.

The major change is that the final `quorumCertificate` is only known at the time of committal, not proposal, so the corresponding changes were made as well as interface changes to make the terminology easier to follow with the corresponding documentation.

**General**
- Update some comments / TODOs throughout the code
- Handle errors where appropriate to support interface changes

**UtilityContext interface changes**
- Change `CommitPersistenceContext()` to `Commit(quorumCert)
- Change `ReleaseContext()` to `Release() error`

**PostgresContext interface changes**
- Removed `quorumCert` from the `SetProposalBlock` method signature
- Changed `Commit()` to `Commit(quorumCert)`
- Renamed `ResetContext` to `Release`
- Replace `AppHash` with `UpdateAppHash` 
- Added `ReleaseWriteContext` to the module level interface

**Persistence Module/Interface changes**
- Reduce the cope of the `StoreBlock` function
- Make `storeBlock` accept a `quorumCert`
- Remove `quorumCertificate` from local state and corresponding setters/getters
- Reduce the cope of the `IndexTransactions` function and remove from interface

**Consensus module changes**
- Add error handling where appropriate corresponding to the interface changes

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] If applicable, I have made corresponding changes to related local or global README
- [ ] If applicable, I have updated the corresponding CHANGELOG
- [ ] If applicable, I have added tests that prove my fix is effective or that my feature works
- [ ] If applicable, I have added new diagrams using [mermaid.js](https://mermaid-js.github.io)

Olshansk committed:
Added updateStateCommitment and GetAppsUpdated

Olshansk committed:
Added celestiaorg/smt tree

Olshansk committed:
Added updating merkle tree using app protos

Olshansk committed:
Merge branch 'main' into issues/147/statehash

Olshansk committed:
WIP - trees

Olshansk committed:
Merge branch 'main' into issues/147/statehash

Olshansk committed:
Update go.mod

Olshansk committed:
Got tests to pass after self review

Olshansk committed:
Merge with main but still not compiling

Olshansk committed:
Resolved some conflicts to make 'make develop_test' work

Olshansk committed:
WIP

Olshansk committed:
Interim commit on block related stuff

Olshansk committed:
Merge branch 'main' into issues/147/statehashv2

Olshansk committed:
Interim commit while cleaning up consensus module

Olshansk committed:
Tests running but not passing

Olshansk committed:
Using proper getters when accessing hotstuff

Olshansk committed:
make develop_test passed

Olshansk committed:
Remove OLSH logs

Olshansk committed:
/s/aggregateMessage/m.aggregateMessage

Olshansk committed:
Updated validateBlockBasic

Olshansk committed:
Removed PersistenceContext exposure from utility context interface

Olshansk committed:
Tests passing - commit before creating a new block type

Olshansk committed:
Added a persistence block proto type

Olshansk committed:
Rename block persistence proto name

Olshansk committed:
Rough non-functional completion of merkle tree flow

Olshansk committed:
Merged with main and resolve conflicts

Olshansk committed:
Remove test_persistence_state_hash

Olshansk committed:
Removed explicit MaxBlockBytes variable

Olshansk committed:
Update block.go

Olshansk committed:
Fix typo in consensus proto

Olshansk committed:
Updated interface to findHighQC

Olshansk committed:
Reverted changes to the utility module to simplify the PR

Olshansk committed:
Reverted changes to the utility module interface

Olshansk committed:
Reverted changes to the persistence module interface

Olshansk committed:
Reverted changes to the persistence module files

Olshansk committed:
Using GetCodec().Marshal() where appropriate

Olshansk committed:
Removed additional persistence related code

Olshansk committed:
Added initialization flow for AppHash

Olshansk committed:
make develop_test passed

Olshansk committed:
Added call to StoreBlock back

Olshansk committed:
Avoiding anypb in a few places in the consensus code

Olshansk committed:
Fixed unit tests

Olshansk committed:
Add back stateHash implementation

Olshansk committed:
Isolate interface changes only

Olshansk committed:
Undo more change

Olshansk committed:
Update the interfaces

Olshansk commented:
@andrewnguyen22 @deblasis Please take a preliminary review at the PR before I continue moving forward with the interface changes (implementation will follow subsequently).

The main thing to inspect is the flow I added. Make sure to view the "rich diff" of the ReadMe.

<img width="1466" alt="Screen Shot 2022-09-26 at 8 54 42 PM" src="https://user-images.githubusercontent.com/1892194/192428320-7b127daf-f981-4f6f-8bb8-5918380dd872.png">

Olshansk committed:
Merge branch 'main' into issues/249/consensus_techdebt

Olshansk committed:
Added shouldElectNextLeader

Olshansk committed:
Added a comment for getSignableBytes

Olshansk committed:
Check for nil in CreateProposeMessage

Olshansk committed:
Revert order of pacemaker defer and telemetry event emitting

Olshansk committed:
s/IdToValAddrMap/idToValAddrMap

Olshansk committed:
s/ValAddrToIdMap/valAddrToIdMap

Olshansk committed:
s/NodeId/nodeId

Olshansk committed:
s/LeaderId/leaderId

Olshansk committed:
s/MessagePool/messagePool

Olshansk committed:
Only a few remaining exported fields in ConsensusModule based on reflections

Olshansk committed:
Add TODO in sequence diagram flow

Olshansk committed:
Added isNodeLockedOnPastQC

Olshansk committed:
Logging error when refreshUtilityContext fails

Olshansk committed:
Added shouldLogHotstuffDiscardMessage

Olshansk committed:
Added shouldPrepareNewBlock

Olshansk committed:
Added TODO for 256

Olshansk committed:
Merge with main

andrewnguyen22 reviewed (commented):
Maybe it's just me, but these mermaid sequence diagrams seem difficult to follow, I'd almost prefer LaTeX code over these diagrams. 

- [x] High cognitive load ( a lot going on in each diagram )
- [x] I fully understand the content of the diagrams, but find it difficult to understand when presented this way
- [x] It's not clear to me what the 'outer' blocks like `Persistence` vs `Persistence (PostgresDB)` are representing in this context
- [x] Mermaid seems like an easy tool to create these

Olshansk commented:
@andrewnguyen22 Appreciate the review, and there's a lot going on in those few bullet points, so I'm going to break it down into a few pieces and going to number it so its easier to respond.

> Maybe it's just me

1. @deblasis Would appreciate your feedback too, especially because you even have less context

2. @andrewnguyen22 Can you think of anyone else (@iajrz?) to provide additional feedback? I've actually been thinking of reaching out to some of the cosmos people who have been thinking about their persistence replacement. Assuming they'd respond (big if, but assuming), wdyt?

> these mermaid sequence diagrams

3. Do you mean [sequence diagrams](https://en.wikipedia.org/wiki/Sequence_diagram) in general or [mermaid's implementation of sequence diagrams](https://mermaid-js.github.io/mermaid/#/sequenceDiagram)?

> I'd almost prefer LaTeX code over these diagrams.

4. Fwiw, latex also has sequence diagrams: https://tex.stackexchange.com/questions/207240/drawing-simple-sequence-diagram.

5. I think sequence diagrams and algorithmic code both have a place, but serve different purposes. Sequence diagrams are good for cross-module (e.g. cross-service communication), where latex is better for the implementation. I plan on using the latter in the implementation component.

> Mermaid seems like an easy tool to create these

Agreed. Best tool I've ever seen in terms of simplicity of use for this purpose.

>  I fully understand the content of the diagrams, but find it difficult to understand when presented this way

6. Noted. Would you be down to take a stab at showing something that would be easier for you to interpret? I only ask because I tried to make this simple but not sure how to make it simpler.

7. From a content correctness perspective at least, 👍 ?

>  It's not clear to me what the 'outer' blocks like Persistence vs Persistence (PostgresDB) are representing in this context

Good feedback. Will update.

>  High cognitive load ( a lot going on in each diagram )

8. Noted. I tried to split it into smaller chunks but clearly it wasn't enough. It's meant to be as difficult as reading a whitepaper or implementation specs, so the expectation is that it'll be more difficult than code, and not something you should ever put on a slide, but approachable enough for someone to reference this when implementing a different module. 

As a reference point, are the following sequence diagrams from [Cosmos' documentation](https://docs.cosmos.network/v0.44/building-modules/msg-services.html) or [AWS' documentation](https://aws.amazon.com/blogs/architecture/sequence-diagrams-enrich-your-understanding-of-distributed-architectures/) equally as hard to understand or similar?

I'm trying to gage if:
8.1 Sequence diagrams are the wrong format for this
8.2 Your personal comfort level / familiarity with sequence diagrams is an issue
8.3 If I should break it down into several smaller sequence diagrams
8.4 If the lack of colors / images contributing to the difficulty

![Screen Shot 2022-10-03 at 4 32 06 PM](https://user-images.githubusercontent.com/1892194/193704150-d7c1ba06-0f02-4afc-be96-b3a4256718e7.png)

![Screen Shot 2022-10-03 at 4 34 02 PM](https://user-images.githubusercontent.com/1892194/193704148-1f859a09-9d11-4b82-9efc-95d7ceba411e.png)

Olshansk committed:
Merge with main

Olshansk committed:
Added SetUtilityMod and made utilityContext private

Olshansk committed:
Updated a comment

Olshansk commented:
@andrewnguyen22 Just wanted to reiterate that I'm still thinking on how to make this simpler and open to suggestions on how to make it easier to understand. I think that if we nail this down in terms of ease of understanding (for the team and others), that's when we'll really make it.

Right now, I'm thinking:
1. Adding colors
2. Adding images
3. Splitting it into smaller sub-flows
4. Adding textual steps below the flow chart for additional context

What I don't want to do but open to being convinced otherwise:
1. Using algorithm definitions (i.e. latex pseudocode) where I believe sequence diagrams are more appropriate 
2. Using flow charts (also available in mermaid) where I believe sequence diagrams are more appropriate

Olshansk committed:
Merge branch 'main' into issues/249/consensus_techdebt

Olshansk committed:
Added a TODO for 283

Olshansk committed:
Merge with issues/249/consensus_techdebt

Olshansk committed:
Revert all changes back to main

Olshansk committed:
Updated base to main for simplicity

Olshansk committed:
Updated shared changelog

Olshansk committed:
Commented out the business logic in storeBlock

Olshansk commented:
@andrewnguyen22 THe key files to focus on in the review are:

- [ ] shared/docs/flows/AppHash.md
- [ ] shared/modules/persistence_module.go
- [ ] shared/modules/utility_module.go
- [ ] shared/CHANGELOG.md

Once I have the 👍 , I can I update the other changelogs as well.

One of the things I'm still thinking through (open to suggestions) is how we should go about with breaking interface changes moving forward. In this PR, I made the corresponding changes in place (minimal but to support the new interface), but in the future, I'm thinking that we can:
1.  `TODO(#ImplementationTicketNumber): Deprecate and will be removed in #XXX `
2.  `TODO(#ImplementationTicketNumber): Placeholder and will be implemented in #XXX`

Wdyt?

andrewnguyen22 reviewed (changes requested):
After reviewing, I still feel the same way.

I don't think it's necessarily your execution of the sequence diagram. I think Cosmos' makes sense while Cloudfront's is less intuitive.

I'd prefer if we found another way to represent the interface interaction and flow - but again am calling out for other's opinions on this matter, because I think this might be personal preference.

andrewnguyen22 commented:
> One of the things I'm still thinking through (open to suggestions) is how we should go about with breaking interface changes moving forward. In this PR, I made the corresponding changes in place (minimal but to support the new interface), but in the future, I'm thinking that we can:
> 
> TODO(#ImplementationTicketNumber): Deprecate and will be removed in #XXX
> TODO(#ImplementationTicketNumber): Placeholder and will be implemented in #XXX
> Wdyt?
> 

This seems fine to me

deblasis commented:
@Olshansk:
> 1. @deblasis Would appreciate your feedback too, especially because you even have less context

@andrewnguyen22: 
> I kindof would like to hear from @deblasis or other @pokt-network/engineering-member on this matter to see if
> 
> > 8.2 Your personal comfort level / familiarity with sequence diagrams is an issue
> 
> is the issue.
> 
> I'll review either way


I think that the question we should ask ourselves is:

**Who is the *Persona* that is going to interact with our codebase and end up reading with interest and curiosity these diagrams?**

That said, some background:

I don't have a strong personal preference for diagrams of LaTeX. Probably due to my career path I am used to **not** having any kind of documentation apart from maybe some arrows on a whiteboard or PostIt notes.

Whenever I see a diagram or pseudocode it's like Christmas for me. Otherwise, brownfield projects have always been like: "here's the git repo, gool luck and don't call me if you have questions".

Looking at the diagrams:

### `AppHash/Block Application`:

I had to google what "alt" means in the sequence diagram.
Most likely it's my ignorance in play here but however I was like, of course, **alternative**.

Then I kept following the arrows spanning the various modules and I feel like it makes sense.

I also have a (probably stupid) question: 

As Replica, why do we validate after storing (WriteOp)? Couldn't this be used to create some form of DoS? 

### `AppHash/Block Commit`:

I would dare to say that it looks like a pretty clear high level representation of a happy path but I can see a lot of things potentially going wrong (what happens when instead of an `<- ok` we get a `ko`?) but as a reader I would imagine to find answers to that in the code.

Maybe in AppHash.md we could clarify, even if perhaps redundant, that these are happy paths. 

#### In a nutshell

I feel like we should put stuff out there and get feedback from the community, at some point, hopefully soon, we'll have plenty of complaints and/or praises about the weirdest things in our codebase and all of that is going to be **gold**.

One useful exercise could be the fact of having sequence diagrams like this and also LaTeX for something else and see how they are received.

Just an idea.

deblasis commented on shared/docs/flows/AppHash.md:
```diff
@@ -0,0 +1,70 @@
+# AppHash
```
https://github.com/pokt-network/pocket/pull/252#issuecomment-1268495211

Olshansk commented:
Just wanted to document an offline discussion I had with @andrewnguyen22 regarding next steps. If you were CCeed in the message below, let me know if you'd be interested in providing feedback once the next version of the sequence diagrams is up.

The main question is: `Do these diagrams provide value or not? We're trying to figure out if this is understandable. Let us know if this makes any sense.`

Goals:
* Easily understand the cross-module flow as a starting point before jumping into the code
* Target technical audience
* Decide how (e.g. types of diagrams) we want to use moving forward for visualizing end-to-end flows

**Improvements we can make to this PR based on the discussion:**
1. Cross-module sequence diagrams should be limited to functions in the module interfaces only
2. These diagrams should not jump in an out of details (e.g. some arrows are simple while others are complex)
3. Split the diagrams into smaller pieces: those with interfaces only and those with details
4. Reduce cognitive load
5. Avoid trying to be fancy
6. Limit participants to 5 and num steps to 8
7. Use consistent participants names/types

**Who we will seek feedback from:**
1. Will look to get a 👍  from the rest of the team, specifically @deblasis , @iajrz and @oten91.  Other core team members have some context, though not complete, and are therefore a great target audience.
2. Will look to get a 👍 from some community members. Some potential people we can reach out to are: @moatus + Node pilot team, @pokt-scan members, @Andrew-Pohl, @PoktBlade @DragonDmoney.

Things we agreed on:
- Mermaid is a sufficient tool of choice (no reason to use Latex unless we write pseudocode algorithms)

Things we have yet to decide:
- Should all end-to-end flows have at least one sequence diagram (supplemented by others if applicable)

Olshansk committed:
Merge with main

Olshansk committed:
Made things work again

Olshansk committed:
Reverted files that shouldn't be part of the interface change

Olshansk committed:
Cleaning up the interface

Olshansk committed:
make develop_test passed

Olshansk committed:
Replace data | ok with result, err_code

Olshansk committed:
Split and update the sequence diagrams

Olshansk commented:
@andrewnguyen22 @deblasis PTAL again when you have a chance. 

I know this is still not perfect, but let me know if the goal of providing enough context to "move to the implementation" is sufficient and clear here.

Olshansk committed:
Update TODO keyword

andrewnguyen22 commented:
I added @iajrz and @oten91 as reviewers because I think it's very important to ensure this is the proper tool for the job and I would very much like second opinions on how well `sequence` diagrams communicate the information to those with less context

Olshansk commented:
For sure, let's get feedback from everyone on the core team before moving forward. Once we incorporate feedback from @iajrz, @oten91 and @deblasis, we can move forward with it.

Community feedback won't be a blocker but something we can integrate in over time.

deblasis commented on shared/docs/flows/AppHash.md:
````diff
@@ -1,40 +1,92 @@
 ```
 
 ## Block Application
 
````
```suggestion

This flow shows how the `leader` and the `replica`s behave in order to apply a `block` and return a `stateHash`.
```

deblasis commented on shared/docs/flows/AppHash.md:
```diff
@@ -1,40 +1,92 @@
+
+    alt as leader
+        C->>+U: GetProposalTransactions(proposer, maxTxBz, [lastVal])
+        U->>U: reap mempool
```
From this diagram it looks like if the `leader` reaps the mempool **every time**.

But as far as I understand, it has to only if there is no quorum for the `Prepare` otherwise it has to use the block in the QC.

I understand it would be quite challenging to represent the two scenarios but maybe it could be another  `alt`

Olshansk replied:
What if I have a `Note` with the text `Get / Prepare Block` and in a separate diagram, we can show the difference?

Olshansk replied:
Resolving per the discussion below.

iajrz commented on shared/docs/flows/AppHash.md:
```diff
@@ -1,40 +1,92 @@
+        C->>+U: GetProposalTransactions(proposer, maxTxBz, [lastVal])
+        U->>U: reap mempool
+        U->>-C: txs
+        Note over C, U: fallthrough to replica behaviour
```
This note seems to imply the "leader" portion is an "opt", not an "alt".

Olshansk replied:
Resolving per the discussion below.

iajrz commented:
The diagrams are very valuable to me. The spatial information helps follow the sequences and interactions in a way that text makes it hard to follow.

iajrz commented on shared/docs/flows/AppHash.md:
```diff
@@ -1,40 +1,92 @@
+    participant U as Utility
+    participant P as Persistence
+
+    alt as leader
+        C->>+U: GetProposalTransactions(proposer, maxTxBz, [lastVal])
+        U->>U: reap mempool
+        U->>-C: txs
+        Note over C, U: fallthrough to replica behaviour
+    else as replica
+        C->>+U: ApplyBlock(height, proposer, txs, lastVals)
+        loop for each operation in tx
+            U->>P: Get*/Set*
+            P->>U: result, err_code
+            U->>U: validation<br>logic
+            U->>P: StoreTransaction(tx)
+            P->>P: store tx<br>locally
+            P->>U: result, err_code
+        end
+        U->>+P: UpdateAppHash()
+        Note over P: Update State Hash
+        P->>-U: stateHash
+        U->>-C: stateHash
+    end
```
There are a few observations to the diagram, and a suggestion to accompany them.

Given [the conversation above](https://github.com/pokt-network/pocket/pull/252/files#r990511395):
* it seems the activity around getting the proposal transactions is optional (when acting as leader) and does not exclude the rest of the sequence. I put the conditional section in an `opt` block instead of an `alt` block.
* I reflected my understanding of @deblasis' comment with regards to the `reap mempool` message by making it mutually exclusive with `get QC`. I don't know where the QC comes from, but reflected my guess that utility might have it.

Other changes: 
* The loop label might have been meant to be `tx in txs` instead of `operations in tx`.
* When actors message themselves into doing work, they should be activated
* Responses should be dotted lines.

```suggestion
    opt if leader
        C->>U: GetProposalTransactions(proposer, maxTxBz, [lastVal])
        activate U
        alt no quorum for Prepare
        U->>U: reap mempool
        activate U
        deactivate U
        else
        U->>U: get QC
        activate U
        deactivate U
        end
        U-->>C: txs
        deactivate U
    end
    C->>+U: ApplyBlock(height, proposer, txs, lastVals)
    loop for each tx in txs
        U->>+P: Get*/Set*
        P-->>-U: result, err_code
        U->>U: Validation logic
        activate U
        deactivate U
        U->>+P: StoreTransaction(tx)
        P->>P: Store tx locally
        activate P
        deactivate P
        P-->>-U: result, err_code
    end
    U->>+P: UpdateAppHash()
    P->>P: Update state hash
    activate P
    deactivate P
    P-->>-U: stateHash
    U-->>-C: stateHash
```

iajrz replied:
[Here's a link](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNqdVMtu2zAQ_JUFT04jI3cdDDQq0BxVOOqlKoqFtLaIUiRDrgK7Qf69S8mvRkZQRAdC2MfM7HClF9W4llSuIj0NZBv6onEbsK8tyOMxsG60R8tQAEYonI1k4xDn-SrlK9ZG836eLVO2pBB15ERT26nGeQa9AUPYUphC6SmWq1WVw1fiMjjvIprHgDZiw1oULPwYpJBBj7vH3f2fDH4YjPwdzc-bM0oqf0YmqC5ihsE6eBpcGHrYuABlIFFK55JqIg-EHnrqvXPmfcyWrkXJxDnolhi-FR-Cs-0lmsAVOfAuvtd36kl-3gr9Z-_N_t645veiI73tOIOzlwKWwcHGeLDROOdHlwS7kwrQ9l_SNNdtOV7Vp7u1HOdUmUQuRyfjYISKQviV1m3mihDqFtPdCuFWNx_x5yhkzS7QxbYseHdzqWm1OhalccQKNGZ_hbC8Slj-93gn74_CKi8jktzAA8ZucdB00DPlIHI6O8lP2be0cyEnEWPrw6lzXJBlMYurTPUUetStfPMvKVYr7qinWuXy2tIGZZZa1fZVSnFgt97bRuUcBsrUMMo8_CJUvpE9ode_LdtK7Q) to a rendered version of this suggestion

deblasis replied:
Nice one, the "replica" behaviour (and the fallback to replica from the leader) looks clearer as well IMHO

andrewnguyen22 replied:
Good feedback @iajrz gives me more confidence moving forward with sequence diagrams in general. Also valuable edits IMO

jessicadaugherty replied:
Nicely done! @andrewnguyen22 do you still want @oten91 eyes on this before we close it out?

andrewnguyen22 replied:
I wouldn't say it's required for my approval. I am awaiting Olshansk's changes / response to Irving's however

andrewnguyen22 replied:
It's important to note - `StoreTransaction()` isn't called until after the `Commit()` phase along side `StoreBlock()`. Please adjust accordingly

Olshansk replied:
@iajrz The suggestion is 💯 !

Olshansk committed:
Resolve merge conflicts

Olshansk committed:
Remove merge conflict file

Olshansk committed:
Update shared/docs/flows/AppHash.md

Co-authored-by: Alessandro De Blasis <alex@deblasis.net>

Olshansk committed:
Update shared/docs/flows/AppHash.md

Co-authored-by: Alessandro De Blasis <alex@deblasis.net>

Olshansk committed:
Update changelog

Olshansk committed:
Update shared/docs/flows/AppHash.md

Co-authored-by: Irving A.J. Rivas Z. <axel.rivas@gmail.com>

Olshansk committed:
Minor changelog update

Olshansk commented on shared/docs/flows/AppHash.md:
```diff
@@ -1,40 +1,92 @@
-        C->>C: Store utilityContext
-        Note over C, PM: See 'Block Application'
+        C->>C: store context<br>locally
+        Note over C, P: See 'Block Application'
```
Resolving in the meantime, but here's the question: https://stackoverflow.com/questions/74103729/adding-hyperlinks-to-notes-in-mermaid-sequence-diagrams

Olshansk committed:
Added activations and deactivations

Olshansk commented:
@iajrz - I added activations in the other diagrams since it added a lot of clarity.

@deblasis - I played around with hyperlinks to no avail. Will wait for this to get answered: https://stackoverflow.com/questions/74103729/adding-hyperlinks-to-notes-in-mermaid-sequence-diagrams

@andrewnguyen22 - `StoreBlock` is getting deprecated to simplify the overall flow. Note that it is not in any of the flows and I updated the changelog + description to make it clearer. `StoreTransaction` stores it in the local postgres context and `Commit` triggers it to create a block using the "persistence specific" block type. In #285 I added a Block proto encapsulated to the persistence module which is store itself.

![Screen Shot 2022-10-17 at 3 35 57 PM](https://user-images.githubusercontent.com/1892194/196296579-4ff1559b-552c-4af9-bfdb-7a59c72a2158.png)

![Screen Shot 2022-10-17 at 3 36 30 PM](https://user-images.githubusercontent.com/1892194/196296589-5c78cc3c-9938-4bb9-bb77-8bb50a502688.png)

![Screen Shot 2022-10-17 at 3 36 07 PM](https://user-images.githubusercontent.com/1892194/196296600-46269d9d-e947-4edf-96f8-dcbc88f9359b.png)

![Screen Shot 2022-10-17 at 3 38 03 PM](https://user-images.githubusercontent.com/1892194/196296806-4fbcc1a8-af0d-443a-9b4f-182a9771084f.png)

deblasis commented on shared/docs/flows/AppHash.md:
```diff
@@ -1,40 +1,122 @@
+- [Block Application](#block-application)
+- [Block Commit](#block-commit)
+
+<!-- See if there's an answer in this question to add links to notes: https://stackoverflow.com/questions/74103729/adding-hyperlinks-to-notes-in-mermaid-sequence-diagrams -->
```
👍

Olshansk replied:
@deblasis Do you think this needs any other changes?

andrewnguyen22 commented:
> @andrewnguyen22 - StoreBlock is getting deprecated to simplify the overall flow. Note that it is not in any of the flows and I updated the changelog + description to make it clearer. StoreTransaction stores it in the local postgres context and Commit triggers it to create a block using the "persistence specific" block type. In https://github.com/pokt-network/pocket/pull/285 I added a Block proto encapsulated to the persistence module which is store itself.

I'm not sure if I agree that StoreBlock is deprecated. This is blocked by consensus around #168

andrewnguyen22 reviewed (changes requested):
> @andrewnguyen22 - StoreBlock is getting deprecated to simplify the overall flow. Note that it is not in any of the flows and I updated the changelog + description to make it clearer. StoreTransaction stores it in the local postgres context and Commit triggers it to create a block using the "persistence specific" block type. In https://github.com/pokt-network/pocket/pull/285 I added a Block proto encapsulated to the persistence module which is store itself.

I'm not sure if I agree that StoreBlock is deprecated. This is blocked by consensus around #168

Olshansk committed:
Merge branch 'main' into issues/251/statehash_interface

Olshansk committed:
Add call to tx exists

Olshansk commented:
@andrewnguyen22 I added a call to `TransactionExists` in the latest commit (https://github.com/pokt-network/pocket/pull/252/commits/4c05a2c10e52938da7fa6824fb38cdf680f37500).

Combined with the details in #315 and offline discussion, are there any more blockers here?

Olshansk committed:
Merge with main

Olshansk committed:
Fix up changelog

Olshansk committed:
Merge branch 'main' into issues/251/statehash_interface

andrewnguyen22 commented:
@Olshansk I think this is blocked by #315 I should have that done this week. 

After - let's re-visit

Olshansk commented:
> @Olshansk I think this is blocked by #315 I should have that done this week.
> 
> After - let's re-visit

I would say that this is a precusor to #315 since it is a definition of the flow / interface while #315 is the actual implementation.

This has been in review for a few weeks now and I belive it is in agreement with the implementation details in #315.

Having agreement here will also enable me to work on #285 in parallel which is the top priority in this iteration.

Olshansk committed:
Merge branch 'main' into issues/251/statehash_interface

Olshansk committed:
Tests passing

andrewnguyen22 commented:
We had a sync meeting yesterday - which confirmed we need to wait for #315 will review after that is unblocked

Olshansk commented:
> We had a sync meeting yesterday - which confirmed we need to wait for #315 will review after that is unblocked

I will also continue working on #285 in parallel knowing that we'll need to make changes and rebase/merge it, but hopefully internal implementation (i.e. the foundation) will remain the same.

Olshansk committed:
Merge branch 'main' into issues/251/statehash_interface

andrewnguyen22 reviewed (changes requested):
#315 is in! 

Resolve conflicts and request a review when this is ready!

Olshansk committed:
Merge with main

Olshansk committed:
WIP

Olshansk committed:
Unit tests passing

Olshansk committed:
Updated some comments

Olshansk committed:
Update most of the external interfaces

Olshansk committed:
Update comment

Olshansk committed:
Self review of most parts of the code

Olshansk committed:
Small commit

Olshansk committed:
First attempt at fixing things

Olshansk committed:
Merged with localnet fixes

Olshansk committed:
Made things kind of work

Olshansk committed:
Make code easier to understand

Olshansk committed:
Improve logging for p2p

Olshansk committed:
Merge branch 'localnet_fixes' into issues/251/statehash_interface

Olshansk committed:
Removed persistence/docs/PROTOCOL_STATE_HASH.md from this commit

Olshansk committed:
Update persistence/module.go

Co-authored-by: Andrew Nguyen <amnguyen1@mail.usf.edu>

Olshansk committed:
Update persistence/module.go

Co-authored-by: Andrew Nguyen <amnguyen1@mail.usf.edu>

Olshansk committed:
Merge branch 'localnet_fixes' into issues/251/statehash_interface

Olshansk committed:
Minor release fixes

Olshansk commented:
> @Olshansk I notice you are still making changes to this branch - would you like me to review this now?

Yes. This is the precursor to the implementation branch and has documentation of the end to end flow.

If we need to make changes based on the feedback in the implementation, we'll do so in the follow up

Olshansk committed:
Merge with main

deblasis committed:
[Utility] Local Proof of Stake CLI -  RPC server [part 2/2] - Issue #112 (#176)

## Description

This PR introduces a simple RPC (JSONRPC) server that will be used to interact with the CLI and/or other clients

## Issue

Part of Issue #112 but we decided to split the work in 2 separate PRs

## Type of change

Please mark the options that are relevant.

- [x] New feature, functionality or library

## List of changes

- Updated node to start an RPC server if enabled via config

## Testing

- [x] `make test_all`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] If applicable, I have made corresponding changes to related local or global README
- [x] If applicable, I have added new diagrams using [mermaid.js](https://mermaid-js.github.io)
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works

* feat(RPC): scaffolding

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* feat(RPC): scaffolding

* feat(RPC): scaffolding

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* fix(RPC): updated config handling

* fix(config.proto): updated timeout type

* feat(RPC): basic RPC server with naive sync TX broadcast

* fix(RPC): fixed HTTP method for Health and Version routes

* style(RPC): format

* refactor(Utility): RPC server RoutesMap for CLI/client use

* refactor(Utility): RPC server exporting RouteKey and RoutesMap

for CLI/client use

* feat(Utility): RPC OpenApi spec and code generation

* refactor(Utility): RPC server refactoring using code generation

RPC server refactoring with code generation from openapi spec

* feat(Utility): Updated RPC spec

* feat(Utility): Regenerated RPC boilerplate and updates

* docs(Utility): RPC and node docs barebones + RPC spec notes

* refactor(Shared): RPC config defaults -changing soon,I'm centralizin

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* fix(Utility): RPC: updated to use test_artifacts defaults

* docs(Utility): RPC and node basic docs

* chore(Utility): fixed versioning schema

* fix(Utility): RPC configs post-merge

* feat(Consensus): configOptions

* feat(P2P): configOptions

* fix(Utility): RPC fix post merge

* fix(Utility): RPC disabled by default because of TECHDEBT

* fix(RPC): test_artifacts in runtime

* fix(go.mod): tidy

* fix(RPC): added imports used in codegen files

* feat(RPC): config proto

* feat(RPC): RPCModule and noopRpcModule

* refactor(Shared): shared.Create -> shared.CreateNode

* docs(RPC): updated code organization post refactoring

* fix(RPC): gitignoring generated files

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* refactor(Consensus): mocks with Return(nil) and not Do(...)

* docs(RPC): updated changelog versions

* fix(Makefile): generate_rpc_openapi

* fix(Makefile): fix for git clone + make develop_test

* Update rpc/v1/openapi.yaml

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update rpc/doc/CHANGELOG.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update rpc/noop_module.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* chore(Shared): added TODOes for ValidateXXX() in modules

* docs(RPC): broadcast_tx_sync summary updated

* Update rpc/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update rpc/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update rpc/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* fix(Makefile): gitignoring generated files breaks tests fix

* feat(Consensus): CurrentRound() and CurrentStep()

* feat(RPC): /v1/consensus/round_state

* feat(RPC): /v1/consensus/round_state handler

* docs(Docs): Adding some more color config + raw_hex_bytes

* chore(Runtime): comment spacing

* Update runtime/docs/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* chore(Consensus): SetBus mock Do(func(modules.Bus) {}) -> Return()

* docs(Pocket): changelog

* Update app/pocket/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update app/pocket/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* docs(RPC): changelog

* docs(RPC): docs TOC

* docs(RPC): Transports -> Endpoints

* feat(Tooling): swagger-ui

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* docs(RPC): swagger ui link to editor and ref to make cmd

* feat(rpcServer): rpcServer is now IntegratableModule

* chore(Shared): tracking TODO (implement validations) #334

* fix(RPC): merge fix

* chore(go.mod): tidy

* docs(RPC): nit

* fix(RPC): int64 on RoundState fields

* refactor(Shared): unexporting XXModuleName

* feat(node): single source of truth for version + overridable

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>
Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
[Utility] Local Proof of Stake CLI - Issue #112 (#169)

## Description
This PR aims at introducing CLI commands relative to the Utility module: 

- Send
- Stake
- EditStake
- Unstake
- Unpause
- ChangeParameter

Consequently, it introduces an RPC server (HTTP) and the ability to "point" the CLI at specific nodes via flags 

Fixes [issue/112](https://github.com/pokt-network/pocket/issues/112)

## Type of change
Please mark the options that are relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- [x] `make test_all`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md)

## Checklist
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] If applicable, I have made corresponding changes to related or global README
- [x] If applicable, I have added added new diagrams using [mermaid.js](https://mermaid-js.github.io)
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works

* fix(config.proto): updated timeout type

* feat(RPC): basic RPC server with naive sync TX broadcast

* fix(RPC): fixed HTTP method for Health and Version routes

* chore(go.mod): tidy

* style(RPC): format

* chore(go.mod): tidy + vendor

* feat(Utility): CLI RPC client function

* feat(Utility): RPC defaults const

* feat(Utility): GetSignBytes

* feat(Utility): CLI utils

* feat(Utility): stake cmd scaffolding

* refactor(Utility): RPC server RoutesMap for CLI/client use

* refactor(Utility): RPC server exporting RouteKey and RoutesMap

for CLI/client use

* feat(Utility): CLI

calling updated QueryRPC to point at route from map

* refactor(Utility): RPC server exporting RouteKey and RoutesMap

for CLI/client use

* chore(Utility): Removed TODO

* fix(Utility): CLI: making use of the pwd flag

* fix(Utility): CLI: code review feedback

* fix(Utility): CLI custodial stake command: OutputAddr = fromAddr

* refactor(Utility): CLI: refactor command bindings

Also added missing functionality in commands other than Stake

* fix(Utility): Fix route after merge

* refactor(Utility): CLI: named return values fix

* test(Utility): CLI: simplified tests for PK parsing from file

* feat(Utility): RPC OpenApi spec and code generation

* refactor(Utility): RPC server refactoring using code generation

RPC server refactoring with code generation from openapi spec

* feat(Utility): RPC OpenApi spec and code generation

* feat(Utility): Updated RPC spec

* feat(Utility): Regenerated RPC boilerplate and updates

* docs(Utility): RPC and node docs barebones + RPC spec notes

* feat(Utility): Updated RPC spec

* docs(Utility): CLI docs barebones

* fix(Utility): CLI: added missing message handling for account

* fix(Shared): fixed RPC config

* feat(Utility): CLI updated to use the generated RPC client

* chore(go.mod): tidy

* refactor(Utility): CLI + RPC models in server.gen.go

* style(Utility): removed redundant struct definition

* docs(README.md): fixed LICENSE reference

* docs(README.md): updated with references to CLI, RPC and Node docs

* refactor(Shared): RPC config defaults -changing soon,I'm centralizin

* refactor(Shared): RPC config defaults -changing soon,I'm centralizin

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* fix(Utility): RPC: updated to use test_artifacts defaults

* refactor(Shared): RPC config defaults -changing soon,I'm centralizin

* feat(Utility): CLI: updated to use test_artifacts default

* docs(Utility): RPC and node basic docs

* docs(Utility): CLI: Changelog

* fix(Utility): CLI: fixed output to user. It shouldn't be logging

* fix(Utility): CLI code review feedback

* style(Utility): code review feedback

* feat(Tooling): Updated makefile commands to better handle codegen

* feat(Tooling): Updated makefile commands to better handle codegen

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* fix(Utility): Fix duplicated models

* chore(Utility): fixed versioning schema

* feat(Utility): CLI documentation generator + first stab at docs

* fix(Utility): CLI specific fixes

* fix(Utility): types fixes

* fix(Utility): RPC configs post-merge

* feat(Consensus): configOptions

* feat(P2P): configOptions

* feat(Utility): CLI: using configOptions to inject PK

* fix(Utility): RPC configs post-merge

* feat(Consensus): configOptions

* feat(P2P): configOptions

* fix(Utility): RPC fix post merge

* fix(Utility): RPC disabled by default because of TECHDEBT

* chore(go.mod): tidy

* fix(go.mod): tidy

* fix(CLI): test_artifacts

* fix(go.mod): tidy

* fix(CLI): updated to use new typesUtil.ActorType

* fix(CLI): runtime based init

* fix(test_artifacts): fix

* fix(CLI): runtime using WithRandomPK()

* fix(CLI): fixed client-only initialization

* fix(Makefile): protogen first

* feat(Utility): GetActorName function (can we autogenerate these?)

* fix(RPC): test_artifacts in runtime

* fix(go.mod): tidy

* fix(CLI): debug commands are now feature flagged

* chore(go.mod): tidy

* chore(CLI): git rm app/pocket/rpc/client.gen.go

* chore(CLI): s/j/tx and s/prepareTx/prepareTxJson

* refactor(shared): converters

* fix(CLI): debug nits

* refactor(CLI): confirmation and credentials

* chore(CLI): removed HACK todo

* fix(Makefile): premature protoc

* fix(RPC): added imports used in codegen files

* docs(RPC): added swagger editor link

* fix(RPC): added imports used in codegen files

* feat(RPC): config proto

* feat(RPC): RPCModule and noopRpcModule

* refactor(Shared): shared.Create -> shared.CreateNode

* refactor(RPC): moved scaffolding into rpc module

* fix(RPC): removed redundant file

* fix(CLI): debug merge fix

* docs(RPC): fixed link after refactoring

* docs(RPC): updated code organization post refactoring

* Update app/client/cli/utils.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* fix(RPC): gitignoring generated files

* style(Persistence): reverting space change

* refactor(Defaults): runtime/defaults package

* chore(CLI): issue handle

* chore(CLI): Issue #310 links

* refactor(CLI): preallocation fix

* style(CLI): oneMillion

* style(CLI): s/RelayChainIDs/relayChainIDs

* feat(Utility): ActorType.GetName()

* chore(Utility): tracking issue #142

* chore(Utility): GetBytes -> GetCanonicalBytes

* chore(CLI): actorCmd -> accountCmd

* docs(CLI): fromAddr note (address currently sourced from pk)

* chore(Runtime): new default value from main

* refactor(CLI): moving utility functions in utils.go

* chore(CLI): NewDebug -> NewDebugCommand

* fix(RPC): gitignoring generated files

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* refactor(Consensus): mocks with Return(nil) and not Do(...)

* docs(RPC): updated changelog versions

* fix(Makefile): generate_rpc_openapi

* fix(Makefile): fix for git clone + make develop_test

* Update rpc/v1/openapi.yaml

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update rpc/doc/CHANGELOG.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update rpc/noop_module.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* chore(Shared): added TODOes for ValidateXXX() in modules

* docs(RPC): broadcast_tx_sync summary updated

* Update rpc/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update rpc/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update rpc/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* fix(Makefile): gitignoring generated files breaks tests fix

* docs(CLI): added todos for exactArgs

* chore(RPC): added types.go

* refactor(RPC): sourcing defaults from defaults not test_artifacts

* feat(CLI): system commands RPC🤝CLI

* feat(Consensus): CurrentRound() and CurrentStep()

* feat(RPC): /v1/consensus/round_state

* feat(RPC): /v1/consensus/round_state handler

* feat(RPC): /v1/consensus/round_state

* refactor(CLI): system command typoes copypastas

* feat(CLI): consensus commands -RoundState and individual ones

* chore(CLI): typo

* docs(CLI): short commands descriptions

* docs(Config): Adding some more color around configuration

* fix(CLI): tx message signing

* feat(CLI): reporting statuscode and body

* fix(Proto): deterministic

* refactor(CLI): prepareTxJSON -> prepareTxBytes

* docs(Docs): Adding some more color config + raw_hex_bytes

* refactor(CLI): Stake command

* fix(CLI): tx message signing

* feat(CLI): reporting statuscode and body

* Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

* feat(Tooling): swagger-ui

* feat(RPC): cors feature flag

* Update utility/types/message.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* chore(Runtime): comment spacing

* docs(CLI): Changelog

* docs(CLI): changelog

* docs(RPC): changelog

* chore(Runtime): comment spacing

* Update runtime/docs/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* chore(Consensus): SetBus mock Do(func(modules.Bus) {}) -> Return()

* refactor(Consensus): mocks with Return(nil) and not Do(...)

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* chore(Consensus): SetBus mock Do(func(modules.Bus) {}) -> Return()

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* docs(Pocket): changelog

* Update app/pocket/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* Update app/pocket/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* docs(RPC): changelog

* docs(RPC): docs TOC

* docs(RPC): Transports -> Endpoints

* feat(Tooling): swagger-ui

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

* docs(RPC): swagger ui link to editor and ref to make cmd

* feat(rpcServer): rpcServer is now IntegratableModule

* chore(Shared): tracking TODO (implement validations) #334

* fix(RPC): merge fix

* chore(go.mod): tidy

* chore(go.mod): tidy

* feat(Tooling): added empty mock_module package for cold start

* docs(RPC): nit

* Update app/client/cli/consensus.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

* docs(CLI): better commands descriptions

* feat(CLI): boldText helper

* style(CLI): nit: real estate

* refactor(CLI): ConsensusState

* docs(CLI): updated docs

* docs(CLI): \n at the end of sentences in Stake command desc.

* docs(CLI): regenerated docs

* fix(RPC): int64 on RoundState fields

* refactor(Shared): unexporting XXModuleName

* feat(node): single source of truth for version + overridable

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>
Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
```diff
@@ -0,0 +1,173 @@
+
+## Context Management
+
+The `Utility` and `Persistence` modules maintain a context (i.e. an ephemeral states) driven by the `Consensus` module that can be released & reverted as a result of various (e.g. lack of Validator consensus) before the state is committed and persisted to disk (i.e. the block is finalized).
```
```suggestion
The `Utility` and `Persistence` modules maintain a context (i.e. an ephemeral states) driven by the `Consensus` module that can be `released & reverted` (i.e. the block is invalid / no Validator Consensus reached) or can be `committed & persisted` to disk (i.e. the block is finalized).
```

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
```diff
@@ -0,0 +1,173 @@
+    participant P as Persistence
+
+    %% Handle New Message
+    B-->>C: HandleMessage(msg)
```
Let's make it explicit that `msg` is `newRoundMsg`

Ditto with `DECIDE`

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
````diff
@@ -0,0 +1,173 @@
+    end
+```
+
+5. The **HotPOKT lifecycle** takes place so Validators achieve consensus (i.e. steps `PRECOMMIT` and `COMMIT`)
````
5 is orphaned between two charts. This is confusing. 

It also implies that `PRECOMMIT` and `COMMIT` are the entirety of the `HotPOKT lifecycle`

Please capitalize `Lifecycle` if you're going to bold it

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
```diff
@@ -0,0 +1,173 @@
+
+5. The **HotPOKT lifecycle** takes place so Validators achieve consensus (i.e. steps `PRECOMMIT` and `COMMIT`)
+6. The `Consensus` module handle the `DECIDE` message
+7. The final `quorumCertificate` is propagated to the `UtilityContext` on Commit
```
Why do we need to use the Utility Module to `passthrough` the `QC` to the `persistenceModule`?

I.E. can 7&8 be combined?

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
```diff
@@ -0,0 +1,173 @@
+6. The `Consensus` module handle the `DECIDE` message
+7. The final `quorumCertificate` is propagated to the `UtilityContext` on Commit
+8. The final `quorumCertificate` is propagated to the `PersistenceContext` on Commit
+9. The persistence module's internal implementation for [Store Block](../../persistence/docs/PROTOCOL_STATE_HASH.md) must execute.
```
**Okay so this is the big one @Olshansk**

*DISCUSS*

We need the `AppHash` during `ApplyBlock` to come to `Consensus` on whether or not the `ProposalBlock` is Valid. So something is wrong here in the ordering as Step 9 (PROTOCOL_STATE_HASH) should have been executed prior to Step 5...

Maybe I'm misunderstanding what you're implying here by `StateHash` but either way - we need to discuss this

andrewnguyen22 replied:
I notice that below:

```markdown
The persistence module's internal implementation for [Updating a State hash] must be triggered
```
You have the lifecyle trigger properly. So I'm not sure why we are `double executing` the StateHash execution... Sorry if I'm misunderstanding what you're saying

Olshansk replied:
We're on the same page. I updated some links and terminology to reflect the confusion. 

Should be clearer now.

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
```diff
@@ -0,0 +1,173 @@
+        deactivate C
+
+        %% Apply Block
+        Note over C, P: See 'Block Application'
```
Omit `See` to be consistent with the chart below's reference to `StoreBlock`

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
````diff
@@ -0,0 +1,173 @@
+- `CreateAndApplyProposalBlock` - Reaps the mempool for a new set of transaction to validate and propose
+
+```mermaid
+graph TD
````
Cool chart! Very readable!

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
```diff
@@ -0,0 +1,173 @@
+    C[Am I the leader?] --> |Yes| D
+    C[Am I the leader?] --> |No| Z
+
+    D[Did I get any prepareQCs?] --> |Find highest valid PrepareQC| E
```
Question:

Does `HighPrepareQC` imply a `LockedBlock`?

andrewnguyen22 replied:
If a proposer dies before they submit the block it's a `locked_qc`. So a `locked_qc` can be a `high_prepare_qc`?

andrewnguyen22 replied:
If not, let's include it in your chart

Olshansk replied:
> So a `locked_qc` can be a `high_prepare_qc`

Yes.  A `lockedQC` may be either `nil` or `prepareQC`, and since we aggregate `prepareQCs`, it may be a `highPrepareQC`. It looks something like this:

| QC        | NewRound | Prepare | Precommit | Commit | Decide    |
| --------- | -------- | ------- | --------- | ------ | --------- |
| prepareQC | nil      | nil     | QC        | QC     | QC -> nil |
| lockedQC  | nil      | nil     | nil       | QC     | QC -> nil |

--- 

I updated the diagram to make it more explicit but not sure if answered the question directly. Lmk

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
```diff
@@ -0,0 +1,173 @@
+    loop for each tx in txs
+        U->>+P: TransactionExists(txHash)
+        P->>-U: true | false
+        opt if tx is not indexed
```
I'm not sure the `opt` is strong enough to show `loop termination` if the `TransactionExists()`

Because ofc if the `TransactionExists()` then the `Block is Invalid` and we can stop `ApplyBlock`

Olshansk replied:
I decided to simply change the return value to `false (does not exist)` and remove `opt` altogether.

It only shows the happy path but much clearer IMO.

andrewnguyen22 commented on shared/docs/PROTOCOL_STATE_HASH.md:
```diff
@@ -0,0 +1,173 @@
+            end
+        end
+    end
+    U->>+P: UpdateAppHash()
```
TODO: Consolidate `AppHash` and `StateHash`

andrewnguyen22 commented on shared/modules/persistence_module.go:
```diff
@@ -35,39 +39,37 @@ type PersistenceRWContext interface {
+// - Use general purpose parameter methods such as `Set(enum_gov_type, ...)` such as `Set(STAKING_ADJUSTMENT, ...)`
+// - Reference: https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_prefer_single_method_interfaces
+
+// TOD (#149): convert address and public key to string from bytes
```
```suggestion
// TODO (#149): convert address and public key to string from bytes
```

andrewnguyen22 replied:
#149 is closed. Update?

andrewnguyen22 reviewed (changes requested):
Some good content - first pass, I have a few discuss items.

One [big one](https://github.com/pokt-network/pocket/pull/252#discussion_r1020457733) in particular.

Olshansk committed:
Updated txResults comment

Jasonyou1995 committed:
Ticket #211: finding unnecessary public functions in the Persistence Module (#320)

Reducing the scope of public functions in the persistence package from public to private for readability purposes

Olshansk committed:
Update shared/docs/PROTOCOL_STATE_HASH.md

Co-authored-by: Andrew Nguyen <amnguyen1@mail.usf.edu>

Olshansk committed:
Remove unnecessary critical sections

Olshansk committed:
Minor updates in documentation

Olshansk committed:
Removed Latest from function names

Olshansk commented on persistence/db.go:
```diff
@@ -94,6 +86,10 @@ func (pg *PostgresContext) ResetContext() error {
+// 1. Can we remove `Latest` from these Setter & Getter methods
+// 2. Can we scope that to this package?
+// 3. Is `context.go` more appropriate for these than `db.go`?
 func (p PostgresContext) GetLatestProposerAddr() []byte {
```
Updated/responded. Resolving this for now in leu of new potential comments.

Olshansk committed:
Updating references documents

Olshansk committed:
Replace true | false with just false to simplify diagram

Olshansk committed:
Merge with main

Olshansk committed:
Updated the consensus changelog

Olshansk committed:
Updated the persistence changelog

Olshansk committed:
Fix numbering in the README

Olshansk committed:
Updated shared/modules changelog

Olshansk committed:
Updated utility changelog

Olshansk committed:
Merge with main

Olshansk committed:
Improve diagram for high vs locked QC

Olshansk commented:
@andrewnguyen22 Updated.

@deblasis Would appreciate your feedback on the interface as well.

deblasis commented on shared/docs/PROTOCOL_STATE_HASH.md:
```diff
@@ -0,0 +1,181 @@
+---
+
+5. The `Consensus` module handles the `DECIDE` message
+6. The final `quorumCertificate` is propagated to the `UtilityContext` & `PersistenceContext` on `Commit`
```
In the issue description and also here, you talk about the "final quorumCertificate" but this exact same language is not used in the [Specification](https://github.com/pokt-network/pocket-network-protocol/tree/main/consensus). Emphasis on the word **final**. By intuition, I assume we are talking about the `CommitQC`. Am I correct?

If I am correct, do you reckon it would be worth it changing the wording accordingly here?

Olshansk replied:
Great point! Updated

deblasis commented on shared/modules/persistence_module.go:
```diff
@@ -123,15 +125,16 @@ type PersistenceReadContext interface {
 	GetBlocksPerSession(height int64) (int, error)
-	GetLatestProposerAddr() []byte
-	GetLatestBlockProtoBytes() []byte
-	GetLatestBlockHash() string
-	GetLatestBlockTxs() [][]byte
+	GetProposerAddr() []byte
+	GetBlockProtoBytes() []byte
+	GetBlockHash() string
```
Any reason why `GetBlockHashAtHeight` returns []byte and `GetBlockHash` returns string?

Olshansk replied:
Just a result of all the string / byte inconsistencies we have throughout the codebase.

In #284, I consolidate it to: `GetBlockHash(height int64) ([]byte, error)`

deblasis reviewed (commented):
It looks good, just a couple of comments/questions

Olshansk committed:
Merge branch 'main' into issues/251/statehash_interface

Olshansk committed:
Rename finalQC to commitQC

deblasis reviewed (approved):
LGTM
